### PR TITLE
fix project creation with robot account

### PR DESCRIPTION
**Pull request by phin1x**
we tried to create projects using a robot account, but the api call always fails with the message "user robot$<name> not found". it turns out that harbor is trying to get the id of the current authorized user when that user is not a "solution user". since a robot account is not a regular user, the api call always aborts with the error "user not found".

this pr fixes the issue by checking if the current authorized user starts with the robot prefix. if thats the case the owner id is set to one.

Related issues: #14145

### CODE REVIEW

**Comment by heww** [comment]
I don't think this is the correct way to assign the creation permission of the project to the robot account.

Please use (`/system/project`, `create`) to check the permission when the security context is not a solution user.

**Comment by phin1x** [comment]
@heww this is not a problem with the actual permission. the rbac check already has passed. at this point the owner of the project is searched. if it is not a solution user, we search for the current authorized user in the user table! but a robot account is not a regular user so the api always repsonds with "user not found". also a robot account does not has a uid, so we have to assign a fix uid (i assigned the admin uid).

File: src/server/v2.0/handler/project.go
Comment (heww) [inline src/server/v2.0/handler/project.go:160]: Please try to use the security context name to check that the request is from a robot account.

**Comment by phin1x** [comment]
@heww i change the code using the security context name for checking if the current user is a robot acocunt

**Comment by phin1x** [comment]
@heww if think is is already done via the rbac system. project level robot accounts have subjects with a "/project/" prefix, system level robot accounts have subjects with a "/system/" prefix. at the moment, you cannot mix project and system level permissions in a robot account, as we discused in other issues (https://github.com/goharbor/harbor/issues/14512#issuecomment-827062680). Since the rbac subject for creating projects is "/system/project", a project level robot account will never be able to create projects.

**Comment by heww** [comment]
> @heww if think is is already done via the rbac system. project level robot accounts have subjects with a "/project/" prefix, system level robot accounts have subjects with a "/system/" prefix. at the moment, you cannot mix project and system level permissions in a robot account, as we discused in other issues ([#14512 (comment)](https://github.com/goharbor/harbor/issues/14512#issuecomment-827062680)). Since the rbac subject for creating projects is "/system/project", a project level robot account will never be able to create projects.

There is an option to allow everyone to create the project, everyone is the default value for this option in the harbor. When this option is everyone, a project-level robot can create a project.

image

**Comment by phin1x** [comment]
i looked in the code and, yes, you are right. now harbor always checks the permissions when creating a project from a robot account.

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/server/v2.0/handler/project.go
Comment (wy65701436) [inline src/server/v2.0/handler/project.go:107]: if !onlyAdmin {
    if it's a project level robot
        return errors.ForbiddenError
}

Comment (phin1x) [reply src/server/v2.0/handler/project.go:107]: is there a method to check, if the robot acocunt scope is system or project level? i didn'd saw something like a getLevel method.

**Comment by phin1x** [comment]
@wy65701436 i updated the check if the robot account is system level

File: src/server/v2.0/handler/project.go
Comment (wy65701436) [inline src/server/v2.0/handler/project.go:166]: Per discussion with maintainers, we should give the owerID to 0 for robot to avoid any confusion for system. Like in the quota page, the `Owner` column will show the wrong message as robot is not a system admin, given it's a system level robot.

Comment (phin1x) [reply src/server/v2.0/handler/project.go:166]: does the user id 0 exists in the database?

Comment (wy65701436) [reply src/server/v2.0/handler/project.go:166]: no, it's not in the databsae. Set to 0 is a short term decision. For the long term, we should conside the usage of owner id the Harbor world wide. Whether to reserve it, or sepreate it into two column, reference and reference_type to support both harbor user and robot account.

**Review by wy65701436 — CHANGES_REQUESTED** [review]
please reslove all the review comments before merging.

File: src/server/v2.0/handler/project.go
Comment (wy65701436) [reply src/server/v2.0/handler/project.go]: @heww is it better to use `Can` to define whether it's a system level robot?

Comment (wy65701436) [reply src/server/v2.0/handler/project.go]: Or here we can leverage the robot name to query robot, there is an attribute to indicate robot scope.

Comment (phin1x) [reply src/server/v2.0/handler/project.go]: the robot context have a isSystemLevel property, but it was not exported. i exported the property with a getter and used that getter.

Comment (wy65701436) [reply src/server/v2.0/handler/project.go]: Another trick way, as the robot name has a pattern, the project level name should be {projectname}+robotname, which is difference with system level robot. And `+` is illegal character when to issue a new robot, so here we can use whether contains `+` to judge.

File: src/server/v2.0/handler/project.go
Comment (phin1x) [reply src/server/v2.0/handler/project.go:166]: are we save to use a user id, which is not in the database. could we get any problems for example if harbor tries to resolve the owner name by the id?

File: src/server/v2.0/handler/project.go
Comment (phin1x) [reply src/server/v2.0/handler/project.go]: imho using a character to identify if a robot is system or project level is not a good idea. if you set the prefix for example to "robot+" the check will fail. the robot security context already has a property which identifies if a robot is system or project level so why not use it?

File: src/server/v2.0/handler/project.go
Comment (phin1x) [reply src/server/v2.0/handler/project.go:168]: updated the code, now the owner id will be 0 for robot accounts

Comment (phin1x) [reply src/server/v2.0/handler/project.go:168]: dco check fixed

**Comment by phin1x** [comment]
@wy65701436 any updates?

**Comment by wy65701436** [comment]
@phin1x can you rebase the PR to resolve the conflict?

**Comment by phin1x** [comment]
@wy65701436 i rebased the pr

**Comment by wy65701436** [comment]
thanks @phin1x , the robot to create project was covered by https://github.com/goharbor/harbor/pull/15461. Now, you can try the latest master, robot can support replication.

**Pull request closed without merging**
